Importer: Make importer step wait for the site information before show the not found error

gabrielcaires opened the pull request:
Fixes DOTCOM-1767
## Proposed Changes
*  Make the component wait to have the site info before showing the not found


## Why are these changes being made?
The step displays a NotFound state while the site data is loading. This PR aims to make the component wait until all requests are completed.


## Testing Instructions
* Create a free site
* Go to `/setup/site-setup/importerWordpress?siteId={YOUR_SITE_ID}&option=content`
*  Check if, after the loading state, you can see the correct importer content [image 1],  instead of the not found state [image 2]

## Expected
![CleanShot 2025-04-01 at 12 43 21@2x](https://github.com/user-attachments/assets/8dc47248-0148-4d33-9a42-a25722b54541)

## Not Expected
![CleanShot 2025-04-01 at 12 49 22@2x](https://github.com/user-attachments/assets/58991db6-19e7-4ad4-b898-f77a9c306ee5)


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

gabrielcaires committed:
Improve site state loading to avoid uncessary notfound states

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/importer/index.tsx:
```diff
@@ -87,6 +83,10 @@ export function withImporterWrapper( Importer: ImporterCompType ) {
 		const stepNavigator = useStepNavigator( flow, navigation, siteId, siteSlug, fromSite );
 		const currentPath = window.location.pathname + window.location.search;
+		const isRequestingAllSites = useSelector( isRequestingSites );
+		const isRequestingCurrentSite = useSelector( ( state ) =>
+			siteId ? isRequestingSite( state, siteId ) : true
+		);
```
I have concerns about this being true when we don't have a `siteId` defined. Could you share more about why you set it up this way?

gabrielcaires replied:
It makes our UI show the loading state while we don't have any further decisions. 
But I don't have strong opinions about.

gabrielcaires replied:
Thinking better, I will set it to false to not mislead the flag, meaning intent

gabrielcaires replied:
Updated here [3547f17](https://github.com/Automattic/wp-calypso/pull/102277/commits/3547f172ccb3a78b66fd206155be46e7c0e0d202)

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/importer/index.tsx:
```diff
@@ -87,6 +83,10 @@ export function withImporterWrapper( Importer: ImporterCompType ) {
 		const fromSiteData = useSelector( getUrlData );
 		const stepNavigator = useStepNavigator( flow, navigation, siteId, siteSlug, fromSite );
 		const currentPath = window.location.pathname + window.location.search;
+		const isRequestingAllSites = useSelector( isRequestingSites );
```
Switching to the `isRequestingSites` selector means that the downstream loading state is taking a lot longer for me, because we're waiting for all the sites to be re-fetched before we proceed. When using the `hasAllSitesList` selector, we're only waiting for the sites to be populated for the first time, not necessarily from the fetch triggered by this page.

To fix this initial issue, I think it may be worth keeping the original logic that checks `hasAllSitesList`, but add in the fetch for the current site. I have created https://github.com/Automattic/wp-calypso/issues/102281 to track the work to fully switch the component tree over to request only the current site.

gabrielcaires replied:
Updated here [ba35b15](https://github.com/Automattic/wp-calypso/pull/102277/commits/ba35b158c7b0e8e30e20050829e03477247d39fe)

daledupreez reviewed (commented):
This is looking good to me, and I haven't reproduced the issue with the new code yet after a few attempts.

I am, however, seeing longer load times for the page as a result of switching to wait for all sites to be (re)loaded. More inline on this topic.

gabrielcaires committed:
Add back hasAllSitesFetched

gabrielcaires committed:
Set loading default state to false when there is no siteId

gabrielcaires committed:
Track error state due to missing

gabrielcaires committed:
Remove uncessary extra condition

site is the most important information, so doesnt matter if we dont have a site but we have site and siteId

fditrapani commented:
Thanks for the fix Gabriel. I tested this on the front-end and it appears to be working.

andres-blanco commented on client/landing/stepper/declarative-flow/internals/steps-repository/importer/index.tsx:
```diff
@@ -97,18 +102,18 @@ export function withImporterWrapper( Importer: ImporterCompType ) {
-		}
+		useEffect( () => {
+			if ( ! isLoading && ! site ) {
+				recordTracksEvent( 'calypso_importer_missing_site', {
```
shouldn't we be using logstash instead of a track event?

gabrielcaires replied:
updated here [1f9f56d](https://github.com/Automattic/wp-calypso/pull/102277/commits/1f9f56d78a0243ff9bae8798d0179d33d928ad9d)

andres-blanco replied:
Thanks, I'll give it a go once calypso.live finish building

gabrielcaires committed:
Use logstash to log error

gabrielcaires committed:
Merge branch 'trunk' into fix/importer-not-found-state

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/importer/index.tsx:
```diff
@@ -87,6 +85,15 @@ export function withImporterWrapper( Importer: ImporterCompType ) {
+			siteId ? isRequestingSite( state, siteId ) : false
+		);
+
+		const isLoading = useMemo( () => {
+			return ! isImporterStatusHydrated || ! hasAllSitesFetched || isRequestingCurrentSite;
+		}, [ isImporterStatusHydrated, hasAllSitesFetched, isRequestingCurrentSite ] );
```
Good call on memoizing this!

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/importer/index.tsx:
```diff
@@ -97,18 +104,26 @@ export function withImporterWrapper( Importer: ImporterCompType ) {
+					},
+				} );
+			}
+		}, [ importer, isLoading, site, siteId ] );
```
Nit: we're using `siteSlug` in the code, so we should list it as a dep.
```suggestion
		}, [ importer, isLoading, site, siteId, siteSlug ] );
```

gabrielcaires replied:
Fixed 8cdbf88f74b6ccd256549534a676ac2c12a4988f

daledupreez commented on client/landing/stepper/declarative-flow/internals/steps-repository/importer/index.tsx:
```diff
@@ -97,18 +104,26 @@ export function withImporterWrapper( Importer: ImporterCompType ) {
+					site_id: siteId,
+					site_slug: siteSlug,
+					properties: {
+						env: config( 'env_id' ),
```
It looks like the `config` you imported is breaking the build.

Maybe we need two aspects to capture the environment?
 * Use the broader config package: `import config from '@automattic/calypso-config';`
 * Also capture `process.env.NODE_ENV`
```suggestion
						env_id: config( 'env_id' ),
						env: process?.env?.NODE_ENV,
```

gabrielcaires replied:
Oh thanks!
Fixed here [444259d](https://github.com/Automattic/wp-calypso/pull/102277/commits/444259d3cfd26dce119db009acfbc1f35016a1aa)

gabrielcaires committed:
Use correct config module

gabrielcaires committed:
Add siteSlug to hook dependency

daledupreez reviewed (approved):
I've retested and the delay is now reduced. I haven't reproduced the issue yet, so I think we should ship this and keep testing, though I am fairly confident this should address the problem.
